pkg/sypgp/sypgp.go

```diff
@@ -487,138 +489,118 @@ func helpAuthentication() (token string, err error) {
-	}
+		if err, ok := err.(*jsonresp.Error); ok && err.Code == http.StatusUnauthorized {
 
-	resp, err := http.DefaultClient.Do(r)
-	if err != nil {
-		return "", err
+			// The request failed with HTTP code unauthorized. Guide user to fix that.
+			authToken, err := helpAuthentication()
+			if err != nil {
+				return fmt.Errorf("could not obtain or install authentication token: %s", err)
+			}
+
+			// Try to pull key from Key Service again with new auth token.
+			c.AuthToken = authToken
+			pd = client.PageDetails{}
+			if keyText, err = c.PKSLookup(context.TODO(), &pd, search, client.OperationIndex, true, false, nil); err != nil {
+				return err
+			}
+		} else if err.Code == http.StatusNotFound {
+			return fmt.Errorf("no matching keys found for fingerprint")
+		} else {
+			return fmt.Errorf("failed to get key: %v", err)
+		}
 	}
-	defer resp.Body.Close()
 
-	// check if error is authentication failure and help user when it's the case
-	if resp.StatusCode == http.StatusUnauthorized {
-		token, err := helpAuthentication()
+	// Print first page of search results.
+	fmt.Print(keyText)
+
+	// Retrieve 2-N pages of search results from Key Service.
+	for pd.Token != "" {
+		resp, err := AskQuestion("\nDisplay more results? [Y/n] ")
 		if err != nil {
-			return "", fmt.Errorf("could not obtain or install authentication token: %s", err)
+			return err
 		}
-		// try request again
-		r, err := doSearchRequest(search, keyserverURI, token)
-		if err != nil {
-			return "", fmt.Errorf("error while preparing http request: %s", err)
+		if resp != "" && resp != "y" && resp != "Y" {
+			break
 		}
-		resp, err = http.DefaultClient.Do(r)
+
+		keyText, err := c.PKSLookup(context.TODO(), &pd, search, client.OperationIndex, true, false, nil)
 		if err != nil {
-			return "", err
+			return err
 		}
-	}
 
-	if resp.StatusCode == http.StatusNotFound {
-		return "", fmt.Errorf("no keys match provided search string")
+		// Print page of search results.
+		fmt.Print(keyText)
 	}
 
-	b, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return "", err
-	}
-
-	return string(b), nil
+	return nil
 }
 
-// doFetchRequest prepares an HKP get request
-func doFetchRequest(fingerprint, keyserverURI, authToken string) (*http.Request, error) {
-	v := url.Values{}
-	v.Set("op", "get")
-	v.Set("options", "mr")
-	v.Set("search", "0x"+fingerprint)
+// FetchPubkey pulls a public key from the Key Service.
+func FetchPubkey(fingerprint, keyserverURI, authToken string, noPrompt bool) (openpgp.EntityList, error) {
 
-	u, err := url.Parse(keyserverURI)
+	// Decode fingerprint and ensure proper length.
+	var fp [20]byte
+	b, err := hex.DecodeString(fingerprint)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to decode fingerprint: %v", err)
 	}
-	u.Path = "pks/lookup"
-	u.RawQuery = v.Encode()
+	if got, want := len(b), len(fp); got != want {
+		return nil, fmt.Errorf("unexpected fingerprint length: %v", len(b))
```

Comment:
Since we're assigning `got, want := [...]`, I think we can just do `return nil, fmt.Errorf("unexpected fingerprint length: %v", got)`

Perhaps even better would be:

`fmt.Errorf("unexpected fingerprint length: %v; expected length: %v", got, want)`

Reply:
Will do.